File: discord-bot/src/llmClient.js
export class LlmUnavailable extends Error {
  constructor(message) {
    super(message);
    this.name = 'LlmUnavailable';
  }
}

// HTTP client for the internal LLM service (POST /chat). Mirrors the shape of
// directoryClient.js: injectable fetchImpl, X-API-Key auth, a single typed
// error. The service's 429/504/502/401 all surface the same way to the bot
// ("try again shortly"), so they collapse into one LlmUnavailable.
export function createLlmClient({ baseUrl, apiKey, fetchImpl = fetch }) {
  const headers = { 'X-API-Key': apiKey, 'Content-Type': 'application/json' };

  return {
    async chat({ messages, system, model, maxTokens = 1024 }) {
      const body = { messages, max_tokens: maxTokens };
      if (system) body.system = system;
      if (model) body.model = model;

      let resp;
      try {
        resp = await fetchImpl(`${baseUrl}/chat`, {
          method: 'POST',
          headers,
          body: JSON.stringify(body),
        });
      } catch {
        throw new LlmUnavailable('network error reaching llm service');
      }
      if (!resp.ok) throw new LlmUnavailable(`llm service returned ${resp.status}`);
      let data;
      try {
        data = await resp.json();
      } catch {
        throw new LlmUnavailable('malformed llm response');
      }
      return { content: data.content, model: data.model, usage: data.usage };
    },
  };
}

File: services/meeting/src/pipeline/llm_client.py
import httpx


class LlmUnavailable(Exception): ...


class LlmClient:
    def __init__(self, base_url: str, api_key: str, timeout_s: float = 60.0):
        self._base = base_url.rstrip("/")
        self._headers = {"X-API-Key": api_key, "Content-Type": "application/json"}
        self._timeout = timeout_s

    def chat(self, *, messages, system=None, model=None, max_tokens=1500) -> str:
        body = {"messages": messages, "max_tokens": max_tokens}
        if system:
            body["system"] = system
        if model:
            body["model"] = model
        try:
            r = httpx.post(
                f"{self._base}/chat",
                json=body,
                headers=self._headers,
                timeout=self._timeout,
            )
            r.raise_for_status()
            return r.json().get("content", "")
        except (httpx.HTTPError, ValueError) as e:
            raise LlmUnavailable(str(e)) from e
